approaches/byte-pair-encoding.md

```diff
@@ -3,5 +3,32 @@ grand_parent: Approaches
 parent: Neural machine translation
 layout: coming_soon
 title: Byte-pair encoding
-description: A data compression algorithm for machine translation
+description: A data compression algorithm for tokenisation for machine translation
 ---
+
+Byte-pair encoding (BPE) is a data compression algorithm where the most common consecutive pair of bytes is replaced by a byte that is not seen in the data.
+It was first introduced in the article titled "_A new Algorithm for Data Compression_" by Philip Gage in the February 1994 edition of C Users Journal.
+
+A very well-known example of byte-pair encoding data compression is, suppose we have a string `aaabdaaabac` which is compressed such that, the most frequent byte pair `aa` is replaced by Z. We now have `ZabdZabac`. 
+The new most frequent byte pair is `ab` so we replace it by Y such that, we now have `ZYdZYac`. 
+Applying recursive byte pair encoding to encode `ZY` as X, we now have `XdXac`. It cannot be compressed further and to decompress the data, replacement in reverse order is performed. 
+
+The use of byte-pair encoding in NLP ensures that the most common words in the vocabulary are presented as single token whereas the rarest words are broken down into subword tokens instead of introducing it as `<UNK>` tokens.
+Byte-pair encoding works well for both character-level and word-level representations and it can manage large corpora.
+It follows greedy approach to find the best possible solution where it goes through every possible option to merge at each iteration by looking at its frequency.
+
+Implementation of byte-pair encoding is slightly modified to fit for subword tokenisation.
+Here, frequently occuring subwords are merged instead of being replaced by another byte as it was in compression.
+For example, let's consider the following unigram `praise`. It is split up into more frequent words and are represented as `['__sow', 'pr', 'ai', 'se', 'd', '__eow']`, where `'__sow'` and `'__eow'` are the markers for start and end of the tokens.
```

Comment:
We can copy the structure of the previous example here and break down the example into steps.